base/subarray.jl

```diff
@@ -532,82 +526,52 @@ end
     end
     Isyms = Any[:(I[$d]) for d = 1:ni]
     exhead, idxs = index_generate(ndims(P), IV, :V, Isyms)
     quote
+        $meta; $bc
         $exhead
-        getindex(V.parent, $(idxs...))
-    end
-end
-@generated function unsafe_getindex{T,N,P,IV,LD}(V::SubArray{T,N,P,IV,LD}, I::Int...)
-    ni = length(I)
-    if ni == 1 && length(IV.parameters) == LD  # linear indexing
-        meta = Expr(:meta, :inline)
-        if iscontiguous(V)
-            return :($meta; unsafe_getindex(V.parent, V.first_index + I[1] - 1))
-        end
-        return :($meta; unsafe_getindex(V.parent, V.first_index + V.stride1*(I[1]-1)))
-    end
-    Isyms = Any[:(I[$d]) for d = 1:ni]
-    exhead, idxs = index_generate(ndims(P), IV, :V, Isyms)
-    quote
-        $exhead
-        unsafe_getindex(V.parent, $(idxs...))
+        @inbounds r = V.parent[$(idxs...)]
+        r
     end
 end
 @generated function setindex!{T,N,P,IV,LD}(V::SubArray{T,N,P,IV,LD}, v, I::Int...)
     ni = length(I)
+    bc = :(@boundscheck checkbounds(V, I...))
+    meta = Expr(:meta, :inline)
     if ni == 1 && length(IV.parameters) == LD  # linear indexing
-        meta = Expr(:meta, :inline)
         if iscontiguous(V)
-            return :($meta; setindex!(V.parent, v, V.first_index + I[1] - 1))
+            return :($meta; $bc; @inbounds V.parent[V.first_index + I[1] - 1] = v; V)
         end
-        return :($meta; setindex!(V.parent, v, V.first_index + V.stride1*(I[1]-1)))
+        return :($meta; $bc; @inbounds V.parent[V.first_index + V.stride1*(I[1]-1)] = v; V)
     end
     Isyms = Any[:(I[$d]) for d = 1:ni]
     exhead, idxs = index_generate(ndims(P), IV, :V, Isyms)
     quote
+        $meta; $bc
         $exhead
-        setindex!(V.parent, v, $(idxs...))
-    end
-end
-@generated function unsafe_setindex!{T,N,P,IV,LD}(V::SubArray{T,N,P,IV,LD}, v, I::Int...)
-    ni = length(I)
-    if ni == 1 && length(IV.parameters) == LD  # linear indexing
-        meta = Expr(:meta, :inline)
-        if iscontiguous(V)
-            return :($meta; unsafe_setindex!(V.parent, v, V.first_index + I[1] - 1))
-        end
-        return :($meta; unsafe_setindex!(V.parent, v, V.first_index + V.stride1*(I[1]-1)))
-    end
-    Isyms = Any[:(I[$d]) for d = 1:ni]
-    exhead, idxs = index_generate(ndims(P), IV, :V, Isyms)
-    quote
-        $exhead
-        unsafe_setindex!(V.parent, v, $(idxs...))
+        @inbounds V.parent[$(idxs...)] = v
+        V
     end
 end
 
 # Indexing with non-scalars. For now, this returns a copy, but changing that
 # is just a matter of deleting the explicit call to copy.
-getindex{T,N,P,IV}(V::SubArray{T,N,P,IV}, I::ViewIndex...) = copy(slice(V, I...))
-unsafe_getindex{T,N,P,IV}(V::SubArray{T,N,P,IV}, I::ViewIndex...) = copy(slice_unsafe(V, I))
+function getindex{T,N,P,IV}(V::SubArray{T,N,P,IV}, I::ViewIndex...)
+    @_inline_meta
+    @boundscheck checkbounds(V, I...)
+    copy(slice_unsafe(V, to_indexes(I...)))
```

Comment:
Yes, [it happens](https://github.com/JuliaLang/julia/blob/c205457df2a00714af4c6d82053ff032cf8684dd/base/subarray.jl#L39-L44) between `slice` and `slice_unsafe`.  This is a better order, though, since it's way faster to check the bounds of a logical vector before you call `find` on it.